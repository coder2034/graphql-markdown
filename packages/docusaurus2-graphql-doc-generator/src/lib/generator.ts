import { promises as fs } from "fs";
import path from "path";

import prettier from "prettier";

import { generateMarkdownFromSchema } from "@edno/graphql-markdown";

import { PluginOptions, Sidebar } from "../types";

const SIDEBAR_FILE = "sidebar-schema.js";

const prettifyJSON = (content: string): string => {
  return prettier.format(content, { parser: "json" });
};

const prettifyMarkdown = (content: string): string => {
  return prettier.format(content, { parser: "markdown" });
};

const renderHomepage = async (
  homepageLocation: string,
  options: PluginOptions
): Promise<string> => {
  const homepage = await fs.readFile(homepageLocation, "utf8");

  const filePath = path.join(options.rootPath, path.basename(homepageLocation));
  await fs.writeFile(filePath, prettifyMarkdown(homepage), "utf8");

  return path.relative("./", filePath);
};

const renderSidebar = async ({
  rootPath,
  baseURL,
}: PluginOptions): Promise<string> => {
  const sidebar: Record<string, Sidebar> = {
    schemaSidebar: [
      {
        dirName: baseURL,
        type: "autogenerated",
      },
    ],
  };

  const filePath = path.join(rootPath, SIDEBAR_FILE);
  await fs.writeFile(filePath, prettifyJSON(JSON.stringify(sidebar)), "utf8");

  return path.relative("./", filePath);
};

export const generateDocFromSchema = async (
  schemaLocation: string,
  options: PluginOptions
): Promise<{ pages: number; sidebar: string }> => {
  const pages = await generateMarkdownFromSchema({ saveToFiles: true });

  // eslint-disable-next-line no-magic-numbers
  if (typeof pages === "undefined" || pages.length === 0) {
    throw new Error("Parsing error, no pages generated");
  }

  if (typeof options === "undefined") {
    throw new Error("Parsing error, no pages generated");
  }

  await renderHomepage(options.homepage, options);
  const sidebarPath = await renderSidebar(options);

  return { pages: pages.length, sidebar: sidebarPath };
};
