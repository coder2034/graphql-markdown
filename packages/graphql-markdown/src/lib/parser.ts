import {
  ASTNode,
  ArgumentNode,
  BooleanValueNode,
  DirectiveDefinitionNode,
  DirectiveNode,
  DocumentNode,
  EnumValueDefinitionNode,
  EnumValueNode,
  FieldDefinitionNode,
  FloatValueNode,
  InputValueDefinitionNode,
  IntValueNode,
  ListTypeNode,
  ListValueNode,
  NameNode,
  NamedTypeNode,
  NonNullTypeNode,
  ObjectFieldNode,
  ObjectTypeDefinitionNode,
  ObjectValueNode,
  OperationTypeDefinitionNode,
  OperationTypeNode,
  StringValueNode,
  TypeDefinitionNode,
  visit,
} from "graphql";

import { ParsedNode } from "..";

const GRAPHQL_KIND = {
  DirectiveDefinition: "DirectiveDefinition",
  EnumTypeDefinition: "EnumTypeDefinition",
  InputObjectTypeDefinition: "InputObjectTypeDefinition",
  InputTypeDefinition: "InputTypeDefinition",
  InterfaceTypeDefinition: "InterfaceTypeDefinition",
  ObjectTypeDefinition: "ObjectTypeDefinition",
  OperationTypeDefinition: "OperationTypeDefinition",
  ScalarTypeDefinition: "ScalarTypeDefinition",
  UnionTypeDefinition: "UnionTypeDefinition",
} as const;

const SIMPLIFIED_NODE_KIND = {
  DirectiveDefinition: "directive",
  EnumTypeDefinition: "enum",
  InputObjectTypeDefinition: "input",
  InputTypeDefinition: "interface",
  Mutation: "mutation",
  ObjectTypeDefinition: "object",
  OperationTypeDefinition: "operation",
  Query: "query",
  ScalarTypeDefinition: "scalar",
  Subscription: "subscription",
  UnionTypeDefinition: "union",
} as const;

const OPERATION_TYPE = {
  MUTATION: "Mutation",
  QUERY: "Query",
  SUBSCRIPTION: "Subscription",
} as const;

const visitor = {
  Argument: (node: ArgumentNode) => {
    return node;
  },
  BooleanValue: ({ value }: BooleanValueNode) => {
    return value;
  },
  Directive: (node: DirectiveNode) => {
    return { ...node, name: `@${node.name}` };
  },
  DirectiveDefinition: (node: DirectiveDefinitionNode) => {
    return { ...node, name: `@${node.name}` };
  },
  Document: ({ definitions }: DocumentNode) => {
    return definitions;
  },
  EnumValue: ({ value }: EnumValueNode) => {
    return value;
  },
  EnumValueDefinition: (node: EnumValueDefinitionNode) => {
    return node;
  },
  FieldDefinition: ({
    type,
    name,
    arguments: args,
    directives,
    kind,
    description,
  }: FieldDefinitionNode) => {
    return {
      ...type,
      arguments: args,
      description,
      directives,
      kind,
      name,
    };
  },
  FloatValue: ({ value }: FloatValueNode) => {
    return value;
  },
  InputValueDefinition: ({
    type,
    name,
    directives,
    defaultValue,
    kind,
    description,
  }: InputValueDefinitionNode) => {
    return {
      ...type,
      defaultValue,
      description,
      directives,
      kind,
      name,
    };
  },
  IntValue: ({ value }: IntValueNode) => {
    return value;
  },
  ListType: ({ type }: ListTypeNode) => {
    return {
      ...type,
      isList: true,
    };
  },
  ListValue: ({ values }: ListValueNode) => {
    return values;
  },
  Name: ({ value }: NameNode) => {
    return value;
  },
  NamedType: ({ name }: NamedTypeNode) => {
    return {
      isList: false,
      isNull: true,
      type: name,
    };
  },
  NonNullType: ({ type }: NonNullTypeNode) => {
    return {
      ...type,
      isNull: false,
    };
  },
  NullValue: () => {
    return "null";
  },
  ObjectField: (node: ObjectFieldNode) => {
    return node;
  },
  ObjectValue: ({ fields }: ObjectValueNode) => {
    return fields;
  },
  OperationType: (node: OperationTypeNode) => {
    return node.toString();
  },
  OperationTypeDefinition: (node: OperationTypeDefinitionNode) => {
    return node;
  },
  StringValue: ({ value }: StringValueNode) => {
    return value;
  },
  TypeDefinition: (node: TypeDefinitionNode) => {
    return node;
  },
} as const;

const OperationTypes: readonly string[] = [
  OPERATION_TYPE.QUERY,
  OPERATION_TYPE.MUTATION,
  OPERATION_TYPE.SUBSCRIPTION,
] as const;

const isNodeTypeOperation = (
  node: ASTNode | ParsedNode
): node is ObjectTypeDefinitionNode => {
  return (
    "name" in node &&
    typeof node.name !== "undefined" &&
    OperationTypes.includes(node.name.toString()) &&
    "fields" in node
  );
};

export const getSimplifiedNodeKind = (node: ASTNode | ParsedNode): string => {
  const kind = "operation" in node ? node.operation : node.kind;
  return SIMPLIFIED_NODE_KIND[kind] ?? node.kind;
};

export const parseSchema = (schema: DocumentNode): ParsedNode[] => {
  const nodes: ASTNode[] = visit(schema, { leave: visitor });
  return nodes.flatMap((node) => {
    if (isNodeTypeOperation(node) && typeof node.fields !== "undefined") {
      return node.fields.map<ParsedNode>((operation: FieldDefinitionNode) => {
        const parsedNode = {
          ...operation,
          kind: GRAPHQL_KIND.OperationTypeDefinition,
          operation: node.name,
        } as unknown as ParsedNode;
        return {
          ...parsedNode,
          simplifiedKind: getSimplifiedNodeKind(parsedNode),
        } as ParsedNode;
      });
    }
    return {
      ...node,
      simplifiedKind: getSimplifiedNodeKind(node),
    } as ParsedNode;
  });
};
