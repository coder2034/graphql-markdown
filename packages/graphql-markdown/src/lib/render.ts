import * as path from "path";
import { promises as fs } from "fs";

import * as Eta from "eta";
import * as prettier from "prettier";
import slugify from "slugify";
import kebabCase from "kebab-case";

import { __basedir, getConfigurationOption, getSimplifiedNodeKind } from "..";

export const renderNode = async (node: any): Promise<string> => {
  const layout = path.resolve(
    __basedir,
    getConfigurationOption("layouts"),
    "layout"
  );
  const result = (await Eta.renderFile(layout, node)) as string;

  return prettier.format(result, { parser: "markdown" });
};

export const slug = (input) => slugify(kebabCase(input));

export const saveMarkdownFile = async (node: any): Promise<string> => {
  const folder = getSimplifiedNodeKind(node);
  const filename = `${slug(node.name)}.md`;
  const filePath = path.resolve(
    __basedir,
    getConfigurationOption("output"),
    slug(folder),
    filename
  );

  const markdown = await renderNode(node);
  await fs.writeFile(filePath, markdown);

  return filePath;
};
