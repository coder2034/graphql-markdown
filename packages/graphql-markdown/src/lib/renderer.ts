import * as path from "path";

import * as Eta from "eta";
import * as prettier from "prettier";

import { __basedir, getConfigurationOption } from "../";

export const renderNode = async (node: any): Promise<string> => {
  const layout = path.resolve(
    __basedir,
    getConfigurationOption("layouts"),
    "layout"
  );
  const result = (await Eta.renderFile(layout, node)) as string;

  return prettier.format(result, { parser: "markdown" });
};
