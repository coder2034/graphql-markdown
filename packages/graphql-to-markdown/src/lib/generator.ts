import * as chalk from "chalk";

import {
  Renderer,
  Printer,
  getSchemaMap,
  loadSchema,
  GraphQLFileLoader,
  UrlLoader,
  JsonFileLoader,
  checkSchemaChanges,
  saveSchemaHash,
  saveSchemaFile,
  round,
} from ".";

const time = process.hrtime();

export async function generateDocFromSchema(
  baseURL: string,
  schemaLocation: string,
  outputDir: string,
  linkRoot: string,
  diffMethod: string,
  tmpDir: string,
): Promise<void> {
  const schema = await loadSchema(schemaLocation, {
    loaders: [new GraphQLFileLoader(), new UrlLoader(), new JsonFileLoader()],
  });

  const renderer: Renderer = new Renderer(
    new Printer(schema, baseURL, linkRoot),
    outputDir,
    baseURL,
  );

  const hasChanges = await checkSchemaChanges(schema, tmpDir, diffMethod);

  if (hasChanges) {
    const rootTypes: any = getSchemaMap(schema);

    const generatedPages = await Promise.all(
      Object.keys(rootTypes).map((typeName) =>
        renderer.renderRootTypes(typeName, rootTypes[typeName]),
      ),
    );

    const pages = generatedPages.reduce(
      (r: any, i: any) => [].concat(r, i),
      [],
    );

    const sidebarPath = renderer.renderSidebar(pages);

    const [sec, msec] = process.hrtime(time);
    const duration = round(sec + msec / 1000000000, 3);
    console.info(
      chalk.green(
        `Documentation successfully generated in "${outputDir}" with base URL "${baseURL}".`,
      ),
    );
    console.log(
      chalk.blue(
        `${pages.length} pages generated in ${duration}s from schema "${schemaLocation}".`,
      ),
    );
    console.info(
      chalk.blue.bold(
        `Remember to update your Docusaurus site's sidebars with "${sidebarPath}".`,
      ),
    );
  } else {
    console.info(
      chalk.blue(`No changes detected in schema "${schemaLocation}".`),
    );
  }
  // create references for checkSchemaChanges
  await saveSchemaHash(schema, tmpDir);
  await saveSchemaFile(schema, tmpDir);
}
