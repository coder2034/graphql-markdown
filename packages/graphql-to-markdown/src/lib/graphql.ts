import {
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLScalarType,
  isListType,
  GraphQLID,
  GraphQLInt,
  GraphQLFloat,
  GraphQLString,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLInputObjectType,
  GraphQLFieldMap,
  GraphQLField,
  GraphQLSchema,
  GraphQLNamedType,
} from "graphql";

import { Maybe } from "graphql/jsutils/Maybe";
import { TypeMap } from "graphql/type/schema";

import { hasOwnMethod, hasOwnProperty } from ".";

export {
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLScalarType,
  isListType,
  GraphQLID,
  GraphQLInt,
  GraphQLFloat,
  GraphQLString,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLInputObjectType,
  isDirective as isDirectiveType,
  getNamedType,
  isScalarType,
  isEnumType,
  isUnionType,
  isInterfaceType,
  isObjectType,
  isInputObjectType as isInputType,
  isNullableType,
  printSchema,
} from "graphql";
export { loadSchema } from "@graphql-tools/load";
export { GraphQLFileLoader } from "@graphql-tools/graphql-file-loader";
export { UrlLoader } from "@graphql-tools/url-loader";
export { JsonFileLoader } from "@graphql-tools/json-file-loader";

export const SCHEMA_EXCLUDE_LIST_PATTERN =
  /^(?!Query$|Mutation$|Subscription$|__.+$).*$/;

export function getDefaultValue(argument: { type: unknown; defaultValue: unknown }) {
  if (isListType(argument.type)) {
    return `[${argument.defaultValue || ""}]`;
  }

  switch (argument.type) {
    case GraphQLID:
    case GraphQLInt:
      return `${argument.defaultValue || "0"}`;
    case GraphQLFloat:
      return `${argument.defaultValue || "0.0"}`;
    case GraphQLString:
    default:
      return argument.defaultValue ? `"${argument.defaultValue}"` : undefined;
  }
}

export function getFilteredTypeMap(
  typeMap: Record<string, GraphQLNamedType>,
  excludeList = SCHEMA_EXCLUDE_LIST_PATTERN,
): Maybe<TypeMap> {
  if (!typeMap) return;
  return Object.keys(typeMap)
    .filter((key) => excludeList.test(key))
    .reduce((res: TypeMap, key) => ((res[key] = typeMap[key]), res), {});
}

export function getIntrospectionFieldsList(queryType: Maybe<GraphQLObjectType<unknown, unknown>>): Maybe<GraphQLFieldMap<unknown, unknown>> {
  if (!queryType){
    return;
  }

  if (!hasOwnMethod(queryType, "getFields")) {
    return;
  }
  return queryType.getFields();
}

export function getFields(type: GraphQLObjectType<unknown, unknown>): Maybe<GraphQLField<any, unknown, Record<string, unknown>>[]> {
  if (!hasOwnMethod(type, "getFields")) {
    return;
  }
  const fieldMap = type.getFields();
  return Object.keys(fieldMap).map((name) => fieldMap[name]);
}

export function getTypeName(
  type: GraphQLObjectType<unknown, unknown>,
  defaultName = "",
): Maybe<string> {
  if (!type) {
    return;
  }
  return (
    (hasOwnProperty(type, "name") && type.name) ||
    (hasOwnMethod(type, "toString") && type.toString()) ||
    defaultName
  );
}

export function getTypeFromTypeMap(typeMap: TypeMap, type: any): Maybe<Record<string, unknown>> {
  if (!typeMap) return;
  return Object.keys(typeMap)
    .filter((key) => typeMap[key] instanceof type)
    .reduce((res: any, key) => ((res[key] = typeMap[key]), res), {});
}

export function getSchemaMap(schema: GraphQLSchema) {
  const typeMap = getFilteredTypeMap(schema.getTypeMap());
  if (!typeMap) return;
  return {
    queries: getIntrospectionFieldsList(
      schema.getQueryType && schema.getQueryType(),
    ),
    mutations: getIntrospectionFieldsList(
      schema.getMutationType && schema.getMutationType(),
    ),
    subscriptions: getIntrospectionFieldsList(
      schema.getSubscriptionType && schema.getSubscriptionType(),
    ),
    directives: schema.getDirectives(),
    objects: getTypeFromTypeMap(typeMap, GraphQLObjectType),
    unions: getTypeFromTypeMap(typeMap, GraphQLUnionType),
    interfaces: getTypeFromTypeMap(typeMap, GraphQLInterfaceType),
    enums: getTypeFromTypeMap(typeMap, GraphQLEnumType),
    inputs: getTypeFromTypeMap(typeMap, GraphQLInputObjectType),
    scalars: getTypeFromTypeMap(typeMap, GraphQLScalarType),
  };
}

export function isParametrizedField(field: GraphQLFieldMap<unknown, unknown> | GraphQLField<unknown, unknown, unknown>) {
  return hasOwnProperty(field, "args") && field.args.length > 0;
}

export function isOperation(query: GraphQLFieldMap<unknown, unknown> | GraphQLField<unknown, unknown, unknown>) {
  return hasOwnProperty(query, "type");
}
