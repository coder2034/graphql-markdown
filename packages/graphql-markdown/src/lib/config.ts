import {
  GraphQLExtensionDeclaration,
  GraphQLProjectConfig,
  loadConfig,
} from "graphql-config";
import { DocumentNode } from "graphql";

import { GraphQLFileLoader } from "@graphql-tools/graphql-file-loader";
import { JsonFileLoader } from "@graphql-tools/json-file-loader";
import { UrlLoader } from "@graphql-tools/url-loader";

import { ExtensionAPI, LoadConfigOptions, Maybe } from "..";

const EXTENSION_NAME = "graphql-markdown" as const;

export const OPTION = {
  LAYOUTS: "layouts",
  MDX: "mdx",
  OUTPUT: "output",
};

export type ConfigurationOptions = {
  excludes?: readonly string[];
  layouts?: string;
  mdx?: boolean;
  output?: string;
};

const defaultOptions: ConfigurationOptions = {
  excludes: [] as const,
  [OPTION.LAYOUTS]: "./layouts" as const,
  [OPTION.OUTPUT]: false as const,
  output: "./output" as const,
} as const;

const setFileLoaderExtension: GraphQLExtensionDeclaration = (
  api: ExtensionAPI
) => {
  [new GraphQLFileLoader(), new JsonFileLoader(), new UrlLoader()].forEach(
    (loader) => {
      return api.loaders.schema.register(loader);
    }
  );
  return { name: EXTENSION_NAME };
};

const loadConfiguration = async (
  options?: LoadConfigOptions
): Promise<Maybe<GraphQLProjectConfig>> => {
  const config = await loadConfig({
    ...options,
    extensions: [setFileLoaderExtension],
  });
  return config?.getDefault();
};

export class Configuration {
  static project: Maybe<GraphQLProjectConfig>;
  static extension: ConfigurationOptions;

  static load = async (options?: LoadConfigOptions): Promise<void> => {
    const configuration = await loadConfiguration(options);
    Configuration.project = configuration;
    Configuration.extension = {
      ...defaultOptions,
      ...Configuration.project?.extension(EXTENSION_NAME),
    };
  };

  static get = (name: string): string => {
    return Configuration.extension[name];
  };

  static set = (name: string, value: string): void => {
    Configuration.extension[name] = value;
  };

  // eslint-disable-next-line require-await
  static schema = async (): Promise<Maybe<DocumentNode>> => {
    return Configuration.project!.getSchema("DocumentNode");
  };
}
