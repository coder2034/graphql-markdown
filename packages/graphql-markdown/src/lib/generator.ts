import { Configuration, ConfigurationOptions } from "./config";
import { LoadConfigOptions, ParsedNode, Result } from "..";
import { renderNode, saveMarkdownFile } from "./render";
import { parseSchema } from "./parser";

const loadNodes = async (
  options?: LoadConfigOptions
): Promise<ParsedNode[]> => {
  await Configuration.load(options);

  const schema = await Configuration.schema();

  if (typeof schema === "undefined") {
    throw new Error("An error occurred while loading the schema");
  }

  return parseSchema(schema);
};

type Options = {
  saveToFiles?: boolean;
  loaderOptions?: LoadConfigOptions;
  generatorOptions?: ConfigurationOptions;
};

export const generateMarkdownFromSchema = async (
  options?: Options
): Promise<Result[]> => {
  const nodes = await loadNodes(options?.loaderOptions);

  const processor =
    options?.saveToFiles === true ? saveMarkdownFile : renderNode;

  const results = await Promise.all(
    nodes.map(async (node: ParsedNode): Promise<Result> => {
      const result = await processor(node, options?.generatorOptions);
      return {
        name: node.name,
        type: node.type,
        /* eslint-disable prettier/prettier */
        ...(options?.saveToFiles === true
          ? { filepath: result }
          : { markdown: result }),
        /* eslint-enable */
      };
    })
  );

  return results;
};
