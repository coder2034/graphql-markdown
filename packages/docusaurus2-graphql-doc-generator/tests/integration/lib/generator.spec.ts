import path from "path";

import dirTree from "directory-tree";
import mock from "mock-fs";

import { generateDocFromSchema } from "../../../src/lib/generator";

describe("lib", () => {
  beforeEach(() => {
    mock(
      {
        __data__: mock.load(path.resolve(__dirname, "../../__data__")),
        assets: {
          "generated.md": "Dummy homepage for tweet.graphql",
        },
        // eslint-disable-next-line camelcase
        node_modules: mock.load(
          path.resolve(__dirname, "../../../node_modules")
        ),
        output: {},
        tmp: {},
      },
      { createCwd: true, createTmp: true }
    );
  });

  afterEach(() => {
    mock.restore();
  });

  describe("renderer", () => {
    describe("generateDocFromSchema()", () => {
      test("generates Markdown document structure from GraphQL schema", async () => {
        expect.hasAssertions();

        await generateDocFromSchema("__data__/tweet.graphql", {
          baseURL: "test",
          homepage: "generated.md",
          linkRoot: "test",
          rootPath: "test",
          schema: "test",
        });

        expect(dirTree("output")).toMatchSnapshot();
        expect(dirTree("tmp")).toMatchSnapshot();
      });
    });
  });
});
