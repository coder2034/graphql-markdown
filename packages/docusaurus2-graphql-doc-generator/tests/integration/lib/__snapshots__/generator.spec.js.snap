// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib renderer generateDocFromSchema() generates Markdown document structure from GraphQL schema 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "name": "directives",
      "path": "output/directives",
      "size": 0,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "enums",
      "path": "output/enums",
      "size": 0,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "inputs",
      "path": "output/inputs",
      "size": 0,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "interfaces",
      "path": "output/interfaces",
      "size": 0,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "mutations",
      "path": "output/mutations",
      "size": 0,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "objects",
      "path": "output/objects",
      "size": 0,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "queries",
      "path": "output/queries",
      "size": 0,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "scalars",
      "path": "output/scalars",
      "size": 0,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "subscriptions",
      "path": "output/subscriptions",
      "size": 0,
      "type": "directory",
    },
    Object {
      "children": Array [],
      "name": "unions",
      "path": "output/unions",
      "size": 0,
      "type": "directory",
    },
  ],
  "name": "output",
  "path": "output",
  "size": 0,
  "type": "directory",
}
`;

exports[`lib renderer generateDocFromSchema() generates Markdown document structure from GraphQL schema 2`] = `
Object {
  "children": Array [
    Object {
      "extension": "",
      "name": ".schema",
      "path": "tmp/.schema",
      "size": 64,
      "type": "file",
    },
    Object {
      "extension": ".graphql",
      "name": "schema.graphql",
      "path": "tmp/schema.graphql",
      "size": 1351,
      "type": "file",
    },
  ],
  "name": "tmp",
  "path": "tmp",
  "size": 1415,
  "type": "directory",
}
`;
